File: src/frontend/src/lib/api-models/terms-and-conditions.ts
import { mapOkResponse } from '@/lib/api-models/error';
import type {
  TermsAndConditions as ApiTermsAndConditions,
  TermsAndConditionsListItem as ApiTermsAndConditionsListItem,
  GetLatestTermsAndConditionsResponse as ApiGetLatestTermsAndConditionsResponse,
  ListTermsAndConditionsResponse as ApiListTermsAndConditionsResponse,
  UpsertTermsAndConditionsDecisionRequest as ApiUpsertTermsAndConditionsDecisionRequest,
  CreateTermsAndConditionsRequest as ApiCreateTermsAndConditionsRequest,
} from '@ssn/backend-api';
import { micromark } from 'micromark';

export type GetLatestTermsAndConditionsResponse = TermsAndConditions | null;

export function mapGetLatestTermsAndConditionForUserResponse(
  res: ApiGetLatestTermsAndConditionsResponse,
): GetLatestTermsAndConditionsResponse {
  const okRes = mapOkResponse(res);

  if (okRes.length === 0) {
    return null;
  }

  return mapTermsAndConditionsResponse(okRes[0]);
}

export type TermsAndConditions = {
  id: string;
  content: string;
  comment: string;
  createdAt: Date;
  hasAccepted: boolean;
};

export function mapTermsAndConditionsResponse(
  res: ApiTermsAndConditions,
): TermsAndConditions {
  return {
    id: res.id,
    content: micromark(res.content),
    comment: res.comment,
    createdAt: new Date(Number(res.created_at / 1_000_000n)),
    hasAccepted: res.has_accepted,
  };
}

export type TermsAndConditionsListItem = {
  id: string;
  content: string;
  comment: string;
  createdAt: Date;
  createdBy: string;
};

export function mapTermsAndConditionsListItemResponse(
  res: ApiTermsAndConditionsListItem,
): TermsAndConditionsListItem {
  return {
    id: res.id,
    content: micromark(res.content),
    comment: res.comment,
    createdAt: new Date(Number(res.created_at / 1_000_000n)),
    createdBy: res.created_by,
  };
}

export function mapListTermsAndConditionsResponse(
  res: ApiListTermsAndConditionsResponse,
): TermsAndConditionsListItem[] {
  const items = mapOkResponse(res);
  return items.map(mapTermsAndConditionsListItemResponse).reverse();
}

export type CreateTermsAndConditionsRequest = {
  content: string;
  comment: string;
};

export function mapCreateTermsAndConditionsRequest(
  req: CreateTermsAndConditionsRequest,
): ApiCreateTermsAndConditionsRequest {
  return {
    content: req.content,
    comment: req.comment,
  };
}

export type UpsertTermsAndConditionsDecisionRequest = {
  termsAndConditionsId: string;
  decisionType: TermsAndConditionsDecisionType;
};

export enum TermsAndConditionsDecisionType {
  Accept = 'Accept',
  Reject = 'Reject',
}

export function mapUpsertTermsAndConditionsDecisionRequest(
  req: UpsertTermsAndConditionsDecisionRequest,
): ApiUpsertTermsAndConditionsDecisionRequest {
  return {
    terms_and_conditions_id: req.termsAndConditionsId,
    decision_type:
      req.decisionType === TermsAndConditionsDecisionType.Accept
        ? { Accept: null }
        : { Reject: null },
  };
}

File: src/backend/src/mapping/terms_and_conditions.rs
use crate::{
    data::{self},
    dto::{
        CreateTermsAndConditionsRequest, GetLatestTermsAndConditionsResponse,
        ListTermsAndConditionsResponse, TermsAndConditions, TermsAndConditionsDecisionType,
        TermsAndConditionsListItem, UpsertTermsAndConditionsDecisionRequest,
    },
};
use canister_utils::{ApiResult, Uuid};

pub fn map_get_latest_terms_and_conditions_response(
    res: Option<(Uuid, data::TermsAndConditions, bool)>,
) -> GetLatestTermsAndConditionsResponse {
    res.map(|(id, res, has_accepted)| TermsAndConditions {
        id: id.to_string(),
        created_at: res.created_at,
        content: res.content,
        comment: res.comment,
        has_accepted,
    })
}

pub fn map_list_terms_and_conditions_response(
    items: Vec<(Uuid, data::TermsAndConditions)>,
) -> ListTermsAndConditionsResponse {
    items
        .into_iter()
        .map(|(id, t)| TermsAndConditionsListItem {
            id: id.to_string(),
            content: t.content,
            comment: t.comment,
            created_at: t.created_at,
            created_by: t.created_by.to_string(),
        })
        .collect()
}

pub fn map_create_terms_and_conditions_decision_request(
    req: UpsertTermsAndConditionsDecisionRequest,
    user_id: Uuid,
    created_at: u64,
) -> ApiResult<data::TermsAndConditionsDecision> {
    Ok(data::TermsAndConditionsDecision {
        terms_and_conditions_id: Uuid::try_from(req.terms_and_conditions_id.as_str())?,
        user_id,
        created_at,
        decision_type: match req.decision_type {
            TermsAndConditionsDecisionType::Accept => data::TermsAndConditionsDecisionType::Accept,
            TermsAndConditionsDecisionType::Reject => data::TermsAndConditionsDecisionType::Reject,
        },
    })
}

pub fn map_create_terms_and_conditions_request(
    req: CreateTermsAndConditionsRequest,
    user_id: Uuid,
    created_at: u64,
) -> ApiResult<data::TermsAndConditions> {
    Ok(data::TermsAndConditions {
        comment: req.comment,
        created_by: user_id,
        created_at,
        content: req.content,
    })
}
